compiler/noirc_evaluator/src/brillig/brillig_ir/codegen_stack.rs

```diff
@@ -1,122 +1,155 @@
-        if let Some(destinations) = movements.get(&current_source) {
-            for destination in destinations {
-                self.perform_movements(movements, *destination);
+        // All sinks and their parents have been processed, remaining nodes are part of a loop
+        // Check if a tail_candidate is a branch to a loop
+        for (entry, free) in tail_candidates {
+            let entry_idx = Self::to_index(&entry);
+            if entry_idx < n && children[entry_idx] == 1 {
+                // Use the branch as the temporary register for the loop
+                self.process_loop(
+                    entry_idx,
+                    &Self::from_index(free),
+                    &mut children,
+                    &sources,
+                    &mut processed,
+                );
             }
-            for destination in destinations {
-                self.mov_instruction(*destination, current_source);
+        }
+        if processed == n {
+            return;
+        }
+        // Now process all the remaining loops with a temporary register.
+        // Allocate one temporary per loop to avoid type confusion when reusing registers,
+        // since different loops may contain values of different types.
+        for i in 0..n {
+            if children[i] == 1 {
+                let src = Self::from_index(i);
+                // Copy the loop entry to a temporary register.
+                // Unfortunately, we cannot use one register for all the loops
+                // when the sources do not have the same type
+                let temp_register = self.registers_mut().allocate_register();
+                self.mov_instruction(temp_register, src);
+                self.process_loop(i, &temp_register, &mut children, &sources, &mut processed);
+                self.deallocate_register(temp_register);
             }
         }
     }
-}
 
-#[derive(Default)]
-struct LoopDetector {
-    visited_sources: HashSet<MemoryAddress>,
-    loops: Vec<im::OrdSet<MemoryAddress>>,
-}
-
-impl LoopDetector {
-    fn collect_loops(&mut self, movements: &HashMap<MemoryAddress, HashSet<MemoryAddress>>) {
-        for source in movements.keys() {
-            self.find_loop_recursive(*source, movements, im::OrdSet::default());
+    /// Generates mov opcodes corresponding to a loop, given a node from the loop (entry)
+    /// and a register not in the loop that contains its value (free)
+    fn process_loop(
+        &mut self,
+        entry: usize,
+        free: &MemoryAddress,
+        children: &mut [usize],
+        source: &[MemoryAddress],
+        processed: &mut usize,
+    ) {
+        let mut current = entry;
+        while Self::to_index(&source[current]) != entry {
+            self.perform_movement(current, source[current], children, processed);
+            current = Self::to_index(&source[current]);
         }
+        self.perform_movement(current, *free, children, processed);
+        children[current] = usize::MAX;
```

Comment:
I wanted to say that perhaps it would make it easier to follow the algorithm if the modification of `children` was closer to where the `while` condition is, rather than happen as a side effect of `perform_movement`, but I see here this line seems to do exactly what `perform_movement` already did, which to set `current` as processed.